### DISCUSSION TITLE
Static library compilation

### DISCUSSION THREAD

**Discussion by johnwason** [tesseract-robotics/tesseract#734] [Ideas]
PyPI is the primary package repository used by Python. It is the source for `pip`, the built in tool that is shipped with Python. Having tesseract available on PyPI would significantly increase the potential user base. An example is opencv, which is downloaded millions of times per month from PyPI.

Packages on PyPI need to be statically linked and have symbol visibility tightly controlled. All dependencies (beyond compiler libraries) must be included in the statically linked module, and only the entry points are exposed. No plugin systems can be used, since the symbols are hidden, the plugins won't be able to find the required symbols in the host libraries.

Currently tesseract is not configured to allow static linking. I tried tinkering with static linking, but found there are too many issues for me to fix in my available time.

These commands helped me get started with the project on Windows:

Install dependencies using vcpkg with triplet `x64-windows-static-md`. (On linux, the triplet `x64-linux` produces static libraries) :

```
git clone https://github.com/microsoft/vcpkg.git
cd vcpkg

vcpkg install --triplet x64-windows-static-md eigen3 tinyxml2 console-bridge assimp urdfdom bullet3 fcl octomap orocos-kdl pcl gtest ompl flann jsoncpp taskflow yaml-cpp boost-dll
```

Build using colcon (installed using pip to system Python install):

```
colcon build --merge-install --event-handlers console_cohesion+ --packages-ignore gtest tesseract_python tesseract_python_viewer --cmake-force-configure --cmake-args -DCMAKE_BUILD_TYPE=Release -DPYTHON_EXECUTABLE=C:/opt/ros/noetic/x64/python.exe -DPYTHON_LIBRARY=C:/opt/ros/noetic/x64/Lib/python38.lib -DINSTALL_OMPL=OFF -DINSTALL_OMPL_TAG=master -DBUILD_IPOPT=OFF -DBUILD_SNOPT=OFF -DINSTALL_FCL=OFF -DINSTALL_BULLET=OFF -DBUILD_SHARED_LIBS=OFF -DCMAKE_TOOLCHAIN_FILE=C:\ws\tesseract_wheel\vcpkg\scripts\buildsystems\vcpkg.cmake -DVCPKG_TARGET_TRIPLET=x64-windows-static-md
```

Once tesseract_planning is functioning as a static library, I can reconfigure tesseract_python to produce a single native module and wheel container that can be uploaded to PyPI.

**Comment by Levi-Armstrong** [comment]
A while back we updated Tesseract to allow this. We explicitly removed `SHARED` from the libraries and add the [following](https://github.com/tesseract-robotics/tesseract/blob/d5a8ee6404209bc09a4e072067e8d8171c1e9700/tesseract_common/cmake/tesseract_macros.cmake#L15-L17) so you can build static. Does this not work or is something else missing?

**Reply by johnwason** [reply]
Static linking requires that all "dependencies of dependencies" be explicitly included in `target_link_libraries`. Some packages correctly set the exports, but a lot like assimp do not. It will take some additional fiddling to get all of the dependencies configured correctly.

**Reply by Levi-Armstrong** [reply]
Okay, let me know if there is anything I can help with.
